### PR TITLE
Refactor data table to use context

### PR DESCRIPTION
**Pull request by raq929**
Aaaand... because I wanted to show off cool things you could do with it using context, it adds a CardView component. 

The table instance is set in the context (multiple tables can use the same TableProvider, allowing for cool interactions between tables that I haven't thought of a real world use case for yet.) 

**NOTE**
Tests are currently all failing, but I wanted review on the code while I work on fixing them up.
There are still some clean-up tasks I have to do as well.

### CODE REVIEW

File: src/DataTable/BulkActions.jsx [hunk 1]
```diff
@@ -24,9 +25,10 @@ const CustomToggle = React.forwardRef(({ children, onClick }, ref) => (
 ));
 
 const BulkActions = ({
-  actions, selectedRows, className, ...rest
+  actions, className, tableName, ...rest
 }) => {
   const { width } = useWindowSize();
+  const { selectedFlatRows, rows } = useContext(TableContext).getTableInstance(tableName);
```
Comment (abutterworth) [inline src/DataTable/BulkActions.jsx:31]: Why afford for multiple tables in a table context?

File: src/DataTable/CardView.jsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  return (
+    <CardGrid
+      columnSizes={columnSizes}
+    >
+      {renderCards()}
+    </CardGrid>
+  );
+};
```
Comment (abutterworth) [inline src/DataTable/CardView.jsx:35]: This is very cool. Riffing on the idea: What if you didn't need this component at all. Could a generic `TableItemsList` with props `as` and `tableItemAs` afford for the regular Table rows by default or:

```
<TableContextProvider>
  <DataTable>
     <TableControlBar>
     <TableItemList as={CardGrid} tableItemAs={MiyazakiCard} />
  </DataTable>
</TableContextProvider>
```

File: src/DataTable/Table.jsx [hunk 3]
```diff
@@ -3,24 +3,33 @@ import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import TableHeaderRow from './TableHeaderRow';
 import TableRow from './TableRow';
+import { useRows } from './hooks';
```
Comment (abutterworth) [inline src/DataTable/Table.jsx:6]: Abstracting this out might not be needed if a generic TableRows (ItemsList) component could be made

File: src/DataTable/TableContext.jsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+export const TableContext = createContext({});
+
+export default function TableData({ children }) {
+  const [tableInstances, setTableInstances] = useState({});
+  const setTableInstance = (tableName, tableInstance) => {
+    if (!tableName) {
+      throw new Error('No table name provided to setTableInstance');
+    }
+    setTableInstances({ ...tableInstances, [tableName]: tableInstance });
+  };
```
Comment (abutterworth) [inline src/DataTable/TableContext.jsx:13]: This whole PR is magic, but this is the one piece I'm lost on. Do you think there's a good use case for having multiple tables inside each other where components would need to be able to render display for their "grandparent" table context?

File: src/DataTable/BulkActions.jsx (hunk 1 above)
Comment (raq929) [reply src/DataTable/BulkActions.jsx:31]: Heh, I think because of an off-hand comment that @adamstankiewicz made once. And I kind of think we'll want tables that interact with each other eventually and it's not hard to do?
Potentially also allows you to maintain table state even if you navigate away from the page and comes back?

File: src/DataTable/CardView.jsx (hunk 2 above)
Comment (raq929) [reply src/DataTable/CardView.jsx:35]: oooo, fancy. Let me think about it / play around with that idea.

File: src/DataTable/BulkActions.jsx (hunk 1 above)
Comment (abutterworth) [reply src/DataTable/BulkActions.jsx:31]: The idea that a table context contains two or more tables seem unintuitive and I think you could save table state or something in either case.

File: src/DataTable/TableContext.jsx (hunk 4 above)
Comment (raq929) [reply src/DataTable/TableContext.jsx:13]: Not exactly, but I could definitely think of ways I might want two different tables to interact together. For instance, we basically have a table within a table here: https://portal.stage.edx.org/laelcooffers/admin/coupons
And I could think of ways in which we might want the state of one table to affect another.
I'm not wedded to this idea, but it does allow you to have multiple tables on one page without multiple providers.

File: src/DataTable/CardView.jsx (hunk 2 above)
Comment (raq929) [reply src/DataTable/CardView.jsx]: I think this is for a subsequent iteration. 😄

File: src/DataTable/BulkActions.jsx [hunk 5]
```diff
@@ -1,10 +1,11 @@
-import React, { useMemo } from 'react';
+import React, { useMemo, useContext } from 'react';
 import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import { MoreVert } from '../../icons';
 import {
   Button, Dropdown, useWindowSize, Icon, breakpoints,
 } from '..';
+import DataTableContext from './TableContext';
```
Comment (abutterworth) [inline src/DataTable/BulkActions.jsx:8]: any reason not match the names?

File: src/DataTable/EmptyTableContent.jsx [hunk 6]
```diff
@@ -1,8 +1,14 @@
-import React from 'react';
+import React, { useContext } from 'react';
 import PropTypes from 'prop-types';
 import classNames from 'classnames';
+import DataTableContext from './TableContext';
 
-const EmptyTable = ({ content, className }) => <div className={classNames('pgn__data-table-empty', className)}>{content}</div>;
+const EmptyTable = ({ content, className }) => {
```
Comment (abutterworth) [inline src/DataTable/EmptyTableContent.jsx:6]: Match filename?

File: src/DataTable/RowStatus.jsx [hunk 7]
```diff
@@ -1,17 +1,23 @@
-import React from 'react';
+import React, { useContext } from 'react';
 import PropTypes from 'prop-types';
+import DataTableContext from './TableContext';
 
-const RowStatus = ({ className, pageSize, itemCount }) => (
-  <div className={className}>Showing {pageSize} of {itemCount}</div>
-);
+const RowStatus = ({ className, itemCount }) => {
+  const instance = useContext(DataTableContext);
```
Comment (abutterworth) [inline src/DataTable/RowStatus.jsx:6]: Leftover from before? Might be clearer to just pull out page and rows

File: src/DataTable/Table.jsx [hunk 8]
```diff
@@ -3,24 +3,33 @@ import PropTypes from 'prop-types';
...
 
 const Table = ({
-  getTableProps, headerGroups, getTableBodyProps, rows, prepareRow, isStriped,
+  isStriped, ...rest
```
Comment (abutterworth) [inline src/DataTable/Table.jsx:9]: An aside: I've used `...attrs` for allowing passthroughs to html elements and `...props` for passthroughts to components. I've seen `...rest` like you have here. Do you have thoughts on a consistent guide for this?

File: src/DataTable/index.jsx [hunk 9]
```diff
@@ -69,64 +71,31 @@ function DataTable({
...
-        isPaginated={isPaginated}
-        previousPage={instance.previousPage}
-        nextPage={instance.nextPage}
-        canNextPage={instance.canNextPage}
-        canPreviousPage={instance.canPreviousPage}
-        pageIndex={instance.state.pageIndex}
-        pageCount={instance.pageCount}
-      />
-    </div>
+    <DataTableContext.Provider value={instance}>
```
Comment (abutterworth) [inline src/DataTable/index.jsx:80]: Where does `instance` come from?

File: www/src/components/CodeBlock.jsx [hunk 10]
```diff
@@ -6,6 +6,7 @@ import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live';
 import * as ParagonReact from '~paragon-react';
 import { Add, AddCircle } from '~paragon-icons';
 import * as FontAwesome from '@fortawesome/free-solid-svg-icons'
+import MiyazakiCard from './MiyazakiCard';
```
Comment (abutterworth) [inline www/src/components/CodeBlock.jsx:9]: Let's move `MiyazakiCard` to an `example-components` directory or something

File: www/src/components/MiyazakiCard.jsx [hunk 11]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Card } from 'react-bootstrap';
```
Comment (abutterworth) [inline www/src/components/MiyazakiCard.jsx:3]:
```suggestion
import { Card } from '~paragon-react';
```

File: src/DataTable/TableControlBar.jsx [hunk 12]
```diff
@@ -1,31 +1,33 @@
...
-  isSelectable, selectedFlatRows, toggleAllRowsSelected, bulkActions,
-  isFilterable, filterNames, pageSize, itemCount, columns, rows,
-  resetAllFilters, className, numBreakoutFilters,
+  bulkActions, itemCount, className, noOfBreakoutFilters, ...rest
 }) => {
-  const bulkActionRows = isSelectable ? selectedFlatRows : rows;
+  const instance = useContext(DataTableContext);
+  const {
+    setFilter,
+  } = instance;
```
Comment (abutterworth) [inline src/DataTable/TableControlBar.jsx:16]:
```suggestion
  const { setFilter } = useContext(DataTableContext);
```

File: src/DataTable/TableControlBar.jsx [hunk 13]
```diff
@@ -34,20 +36,14 @@ const TableControlBar = ({
...
           <SmartStatus
-            isSelectable={isSelectable}
-            isFilterable={isFilterable}
-            numberOfSelectedRows={selectedFlatRows.length}
-            toggleAllRowsSelected={toggleAllRowsSelected}
-            filterNames={filterNames}
-            pageSize={pageSize}
             itemCount={itemCount}
-            resetAllFilters={resetAllFilters}
+            {...rest}
```
Comment (abutterworth) [inline src/DataTable/TableControlBar.jsx:40]: What's getting passed to all of these components? (`BulkActions` etc)

File: src/DataTable/index.jsx [hunk 14]
```diff
@@ -20,15 +20,17 @@ import TableHeaderCell from './TableHeaderCell';
...
-  columns, data, bulkActions, defaultColumnValues, additionalColumns, isSelectable,
+  columns, data, defaultColumnValues, additionalColumns, isSelectable,
   isPaginated, manualPagination, pageCount, itemCount,
   isFilterable, manualFilters, fetchData, initialState,
   isSortable, manualSortBy,
   initialTableOptions,
   EmptyTableComponent,
-  numBreakoutFilters,
+  children,
+  ...rest
```
Comment (abutterworth) [inline src/DataTable/index.jsx:33]: If DataTable consumed a DataTableContext, then you could reduce these props to ones mostly about display and composition of children.

File: www/src/pages/table/datatable.mdx [hunk 15]
```diff
@@ -18,6 +18,20 @@ import SingleComponentStatus from '../../components/SingleComponentStatus';
...
 
+## How children get information
+
+## TableContext
+The table context gets the current `react-table` instance of the table from the `DataTable` component and makes it available to any component within the provider,
+whether or not that component is a child of DataTable. It can store instances of multiple tables that are accessed using the table's name.
+
+```jsx
+const instance = useContext(DataTableContext)
+```
```
Comment (abutterworth) [inline www/src/pages/table/datatable.mdx:29]: Inconsistent naming for TableContext

File: www/src/pages/table/datatable.mdx [hunk 16]
```diff
@@ -18,6 +18,20 @@ import SingleComponentStatus from '../../components/SingleComponentStatus';
...
+
+## Props passed to children
+The DataTable will pass the `itemCount` and any extra props on to it's children. Component's outside the `DataTable` component will
```
Comment (abutterworth) [inline www/src/pages/table/datatable.mdx:32]: Why itemCount? Is this not known in the TableContext?

File: src/DataTable/BulkActions.jsx (hunk 5 above)
Comment (raq929) [reply src/DataTable/BulkActions.jsx:8]: Noooo, they should definitely match

File: src/DataTable/RowStatus.jsx (hunk 7 above)
Comment (raq929) [reply src/DataTable/RowStatus.jsx:6]: Oh, it's probably left over from debugging; will fix.

File: src/DataTable/Table.jsx (hunk 8 above)
Comment (raq929) [reply src/DataTable/Table.jsx:9]: Yeah, I've seen all kinds of things. I'm happy to go with `...props` so it's consistent across this repo at least.

File: src/DataTable/TableControlBar.jsx (hunk 13 above)
Comment (raq929) [reply src/DataTable/TableControlBar.jsx:40]: bulkActions, mostly, to maintain the ability to define that on DataTable. I think my idea was to allow people to pass whatever props they wanted down to these components. I don't know how useful that actually is, though.

File: src/DataTable/index.jsx (hunk 9 above)
Comment (raq929) [reply src/DataTable/index.jsx:80]: It comes from `useTable` (from `react-table`). The child components all deal with it not being there gracefully. (I mean, as gracefully as possible)

File: www/src/components/CodeBlock.jsx (hunk 10 above)
Comment (raq929) [reply www/src/components/CodeBlock.jsx:9]: YES good call. I also considered showing it's code in the docs?

File: www/src/pages/table/tablefilters.mdx [hunk 17]
```diff
@@ -53,6 +53,7 @@ The `TextFilter` and `DropdownFilter` components can be used with a variety of t
 The checkbox filter is used in the dropdown.
 
 ```jsx live
+<TableContextProvider >
```
Comment (raq929) [inline www/src/pages/table/tablefilters.mdx:56]: Ooops these need to go away now.

File: src/DataTable/index.jsx [hunk 18]
```diff
@@ -69,64 +71,31 @@ function DataTable({
...
-    getTableBodyProps,
-    headerGroups,
-    rows,
-    prepareRow,
-  } = instance;
-
-  const filterNames = instance.state.filters ? instance.state.filters.map((filter) => filter.id) : [];
-  const resetAllFilters = instance.setAllFilters ? () => instance.setAllFilters([]) : null;
-  const pageSize = instance.page ? instance.page.length : rows.length;
+  const propsForChildren = { itemCount, ...rest };
```
Comment (raq929) [inline src/DataTable/index.jsx:74]: Augment instance with these props.

File: www/src/pages/table/datatable.mdx [hunk 19]
```diff
@@ -18,6 +18,20 @@ import SingleComponentStatus from '../../components/SingleComponentStatus';
 The DataTable component is a wrapper that uses the <a href="https://react-table.tanstack.com/docs">react-table</a> library to
 create tables. It can be used as is, or its subcomponents can be used on their own, allowing the developer full control.
 
+## How children get information
+
+## TableContext
+The table context gets the current `react-table` instance of the table from the `DataTable` component and makes it available to any component within the provider,
+whether or not that component is a child of DataTable. It can store instances of multiple tables that are accessed using the table's name.
+
```
Comment (raq929) [inline www/src/pages/table/datatable.mdx:26]: Add something about `...rest` getting passed into the context.

File: src/DataTable/Table.jsx (hunk 3 above)
Comment (raq929) [reply src/DataTable/Table.jsx:6]: Also going to go on my list for later improvements. 😉

File: src/DataTable/DataTable.scss [hunk 20]
```diff
@@ -1,10 +1,10 @@
...
-$data-table-padding-x: 14px;
-$data-table-padding-y: 14px;
-$data-table-padding-small: 8px;
-$data-table-cell-padding: 8px 14px;
-$data-table-footer-position: flex-end;
+$data-table-padding-x: 14px!default;
+$data-table-padding-y: 14px!default;
+$data-table-padding-small: 8px!default;
+$data-table-cell-padding: 8px 14px!default;
+$data-table-footer-position: center!default;
```
Comment (abutterworth) [inline src/DataTable/DataTable.scss:7]:
```suggestion
$data-table-padding-x: 14px !default;
$data-table-padding-y: 14px !default;
$data-table-padding-small: 8px !default;
$data-table-cell-padding: 8px 14px !default;
$data-table-footer-position: center !default;
```

File: src/DataTable/DataTable.scss [hunk 21]
```diff
@@ -123,3 +123,7 @@ $data-table-footer-position: flex-end;
     margin-top: .36rem;
   }
 }
+
+.pgn__data-table-card-view {
+  padding: 0 $data-table-padding-x
```
Comment (abutterworth) [inline src/DataTable/DataTable.scss:128]:
```suggestion
  padding: 0 $data-table-padding-x;
```

File: src/DataTable/TablePagination.jsx [hunk 22]
```diff
@@ -1,35 +1,41 @@
...
+          {pageIndex + 1} of {pageCount}
+        </strong>{' '}
+      </div>
+      <ButtonGroup>
+        <Button variant="outline-primary" onClick={() => previousPage()} disabled={!canPreviousPage}>
+          {PREVOUS_BUTTON_TEXT}
+        </Button>
+        <Button variant="outline-primary" onClick={() => nextPage()} disabled={!canNextPage}>
+          {NEXT_BUTTON_TEXT}
+        </Button>
```
Comment (abutterworth) [inline src/DataTable/TablePagination.jsx:35]:
```suggestion
        <Button variant="outline-primary" onClick={previousPage} disabled={!canPreviousPage}>
          {PREVOUS_BUTTON_TEXT}
        </Button>
        <Button variant="outline-primary" onClick={nextPage} disabled={!canNextPage}>
          {NEXT_BUTTON_TEXT}
        </Button>
```

File: src/DataTable/TableRow.jsx [hunk 23]
```diff
@@ -26,7 +26,7 @@ TableRow.propTypes = {
   /** props to include on the tr tag (must include id) */
   getRowProps: PropTypes.func.isRequired,
   /** cells in the row */
-  cells: PropTypes.arrayOf(PropTypes.shape()).isRequired,
+  cells: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
```
Comment (abutterworth) [inline src/DataTable/TableRow.jsx:29]: Is `PropTypes.object` more appropriate?

File: src/DataTable/index.jsx [hunk 24]
```diff
@@ -69,64 +72,34 @@ function DataTable({
...
+        {children}
+        {!children
+          && (
+          <>
+            <TableControlBar />
+            <Table />
+            <EmptyTableComponent />
+            <TableFooter />
+          </>
+          )}
```
Comment (abutterworth) [inline src/DataTable/index.jsx:95]: Any reason not to prefer a ternary operator?
```suggestion
        {children ? children : (
          <>
            <TableControlBar />
            <Table />
            <EmptyTableComponent />
            <TableFooter />
          </>
        )}
```

File: src/DataTable/tests/DataTable.test.jsx [hunk 25]
```diff
@@ -101,8 +113,48 @@ describe('<DataTable />', () => {
...
+    expect(spy.mock.calls[0][0].columns).toEqual(props.columns);
+    expect(spy.mock.calls[0][0].data).toEqual(props.data);
+    expect(spy.mock.calls[0][0].initialState).toEqual({});
+    expect(spy.mock.calls[0]).toHaveLength(2);
+  });
+  test.each([
+    [{}, { manualFilters: false, manualPagination: false, manualSortBy: false }],
+    [{ manualFilters: true }, { manualFilters: true, manualPagination: false, manualSortBy: false }],
+    // [{ manualPagination: true }, { manualFilters: true, manualPagination: true, manualSortBy: false }],
+    // [{ manualSortBy: true }, { manualFilters: true, manualPagination: true, manualSortBy: true }],
```
Comment (abutterworth) [inline src/DataTable/tests/DataTable.test.jsx:141]: Why are these here?

File: src/DataTable/DataTable.scss (hunk 21 above)
Comment (raq929) [reply src/DataTable/DataTable.scss:128]: ooooh, we should add a css linter.

**Review by abutterworth — APPROVED** [review]
🚀 🚀 🚀

File: src/DataTable/TableRow.jsx (hunk 23 above)
Comment (raq929) [reply src/DataTable/TableRow.jsx:29]: I'm pretty sure the linter yells at me every time I use `PropTypes.object`?

File: src/DataTable/index.jsx (hunk 24 above)
Comment (raq929) [reply src/DataTable/index.jsx:95]: Ternery is great here; the way it is is an artifact of something I used to be doing but changed.

File: src/DataTable/tests/DataTable.test.jsx (hunk 25 above)
Comment (raq929) [reply src/DataTable/tests/DataTable.test.jsx:141]: Oh, because I was debugging the tests and didn't want to run them all. I'll turn them back on.
